mopidy_local/storage.py

```diff
@@ -1,13 +1,239 @@
+    # original source: http://goo.gl/6bo5Vx
+    index = 0
+    ftype = 0
+    size = 2
+    while not 0xC0 <= ftype <= 0xCF:
+        index += size
+        ftype = data[index]
+        while ftype == 0xFF:
+            index += 1
+            ftype = data[index]
+        index += 1
+        size = struct.unpack(">H", data[index : index + 2])[0] - 2
+        index += 2
+    index += 1  # skip precision byte
+    height, width = struct.unpack(">HH", data[index : index + 4])
+    return width, height
+
+
+class LocalStorageProvider:
+    def __init__(self, config):
+        self._config = ext_config = config[Extension.ext_name]
+        self._media_dir = ext_config["media_dir"]
+        self._data_dir = Extension.get_data_dir(config)
+        self._image_dir = Extension.get_image_dir(config)
+        self._base_uri = "/" + Extension.ext_name + "/"
+        self._patterns = list(map(str, ext_config["album_art_files"]))
+        self._dbpath = self._data_dir / "library.db"
+        self._connection = None
+
+    def load(self):
+        with self._connect() as connection:
+            version = schema.load(connection)
+            logger.debug("Using SQLite database schema v%s", version)
+            return schema.count_tracks(connection)
+
+    def begin(self):
+        return schema.tracks(self._connect())
+
+    def add(self, track, tags=None, duration=None):
+        logger.info("Adding track: %s", track)
+        images = None
+        if track.album and track.album.name:  # FIXME: album required
+            uri = translator.local_uri_to_file_uri(track.uri, self._media_dir)
+            try:
+                images = self._extract_images(track.uri, tags)
+                logger.debug("%s images: %s", track.uri, images)
+            except Exception as e:
+                logger.warn("Error extracting images for %s: %s", uri, e)
+        try:
+            track = self._validate_track(track)
+            schema.insert_track(self._connect(), track, images)
+        except Exception as e:
+            logger.warn("Skipped %s: %s", track.uri, e)
+
+    def remove(self, uri):
+        schema.delete_track(self._connect(), uri)
+
+    def flush(self):
+        if not self._connection:
+            return False
+        self._connection.commit()
+        return True
+
+    def close(self):
+        if self._connection:
+            schema.cleanup(self._connection)
+            self._connection.commit()
+            self._connection.close()
+            self._connection = None
+        else:
+            logger.error("Attempting to close while not connected")
+        self._cleanup_images()
+
+    def clear(self):
+        logger.info("Clearing image directory")
+        try:
+            for root, dirs, files in os.walk(self._image_dir, topdown=False):
+                for name in dirs:
+                    os.rmdir(os.path.join(root, name))
+                for name in files:
+                    os.remove(os.path.join(root, name))
```

Comment:
Could this be replaced with [`shutil.rmtree()`](https://docs.python.org/3/library/shutil.html#shutil.rmtree)?